### PR TITLE
Directly use InputStream for parsing response JSON.

### PR DESCRIPTION
**Pull request by findgriffin**
### Description
This change is a big refactor and the main party trick is that we can use `HttpResponse<InputStream` and then pass that to FaunaParser instead of having to do `toString -> ObjectMapper.read(..)`.

Some other things are mixed in here like some follow-on fixes from the previous PR #143  and the QueryTags `parse` method was incorrectly expecting an array, I've implemented an E2E test to verify that the `a=b,c=d` format is what we really get.

I expect to have a follow-on PR that will completely remove ConstraintFailureWire, ErrorInfoWire and QueryResponseWire, associated dead code, and refactor the tests. I stopped short of doing that in this PR because a lot of tests need to be changed.

### Motivation and context

I started out with the goal of removing ObjectMapper and the associated `*Wire` classes (ConstraintFailureWire, ErrorInfoWire, and QueryResponseWire) and I didn't get all the way there, but this PR is already pretty big.

### How was the change tested?
Many modified unit tests, and I added some E2E tests.

### Change types

* - [x] Bug fix (non-breaking change that fixes an issue)
* - [ ] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [x] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/fauna/response/QueryStats.java [hunk 1]
```diff
@@ -74,43 +75,47 @@ public static QueryStats parseStats(JsonParser parser) throws IOException {
...
+                    default:
+                        throw new ClientResponseException("Unknown field " + fieldName);
                 }
             }
             return new QueryStats(computeOps, readOps, writeOps, queryTimeMs, processingTimeMs, contentionRetries, storageBytesRead, storageBytesWrite, rateLimitsHit);
         } else if (parser.nextToken() == VALUE_NULL) {
             return null;
         } else {
-            throw new ClientException("Query stats should be an object or null");
+            throw new ClientException("Query stats should be an object or null, not " + parser.getCurrentToken());
```
Comment (pnwpedro) [inline src/main/java/com/fauna/response/QueryStats.java:118]:
```suggestion
            throw new ClientResponseException("Query stats should be an object or null, not " + parser.getCurrentToken());
```

File: src/main/java/com/fauna/response/QueryTags.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+public class QueryTags extends HashMap<String, String> {
+
+    /**
+     * Build a set of query tags from a collection of strings.
+     * @param tags   1 or more strings of the form k=v where k is the tag key, and v is the tag value.
+     * @return          A new QueryTags instance.
+     */
+    public static QueryTags of(String... tags) {
+        QueryTags queryTags = new QueryTags();
+        new QueryTags();
```
Comment (pnwpedro) [inline src/main/java/com/fauna/response/QueryTags.java:22]:
```suggestion
```

File: src/test/java/com/fauna/codec/json/QueryTagsParsingTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    @Test
+    public void deserializeEmptyString() throws IOException {
+        JsonParser parser = MAPPER.createParser("\"\"");
+        QueryTags tags = QueryTags.parse(parser);
+        assertTrue(tags.isEmpty());
+    }
+
+    @Test
+    public void deserializeQueryTags() throws IOException {
+        JsonParser parser = MAPPER.createParser("\"foo=bar, baz=foo\"");
```
Comment (pnwpedro) [inline src/test/java/com/fauna/codec/json/QueryTagsParsingTest.java:41]: Is there a space between pairs on the wire?

**Review by pnwpedro — CHANGES_REQUESTED** [review]
A few required and a few optional pieces of feedback.

In the desc, you say:
> QueryTags parse method was incorrectly expecting an array

I don't see a parse method in main. AFAICT, the QueryTags deserializer was doing the right thing expecting a comma-delimited key-value pair syntax (`foo=bar,baz=moo`)

File: src/test/java/com/fauna/codec/json/QueryTagsParsingTest.java (hunk 3 above)
Comment (findgriffin) [reply src/test/java/com/fauna/codec/json/QueryTagsParsingTest.java:41]: I don't know if the server sends back the exact string it got, or parses then re-encodes without spaces, either way, I just wanted to make sure we handled spaces.

**Review by pnwpedro — APPROVED** [review]

**Pull request merged**